src/cmd/helm.rs

```diff
@@ -533,6 +546,9 @@ impl Helm {
             }
             Err(err) => {
                 error!("Helm error: {:?}", err);
+                if let Err(e) = delete_unused_chart_backup(&self.kubernetes_config, &self.get_all_envs(envs), chart) {
+                    return Err(e);
```

Comment:
- You don't check if need_backup was true, if it was not the get_secret in the delete is going to fail

- You return the error of to delete_unsued, while we are interested in the upgrade error, it is going to be confusing in log/to troubleshoot.

I was not very found of putting the logic of backup/restore in the helm upgrade at first, because this is nothing to with the helm command at all.
But the more you add logic of this in this function, you more we see that it tries to mix 2 different things in one.

Just extract the logic of backup/restore outside this helm function and call it above

if Err(err) = helm_utils::backup_ressources(xxxx) {
     return Err(err);
}

match helm::upgrade(xxxx) {
 Ok(x) =>  {
    helm_utils::restore(xxxx); // rollback helm if it fails ?
    Ok(x)
 } 
 Err(x) { helm_utils::delete_backup(); xxxx } 
}

it should be your prepare_backup function that should do the check of versions